presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -739,6 +741,12 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
             List<Expression> orderByExpressions = emptyList();
             if (node.getOrderBy().isPresent()) {
                 orderByExpressions = analyzeOrderBy(node, getSortItemsFromOrderBy(node.getOrderBy()), queryBodyScope);
+
+                if (queryBodyScope.getOuterQueryParent().isPresent() && !node.getLimit().isPresent() && !node.getOffset().isPresent()) {
+                    // not the root scope and ORDER BY is ineffective
+                    analysis.markRedundantOrderBy(node.getOrderBy().get());
+                    warningCollector.add(new PrestoWarning(REDUNDANT_ORDER_BY, "ORDER BY in subquery has no effect"));
```

Comment:
"has no effect" is true only when the feature is on. When the feature is off, the ORDER BY may have user-visible effect, so the warning message should be different

Reply:
Also, warnings are currently hardly visible to users -- only CLI displays them. Let's talk about this.

Reply:
This is meant to convey that the construct has no semantic effect, so users shouldn’t rely on it

Reply:
Warnings are available via JDBC and the UI, if I recall.

Reply:
> Warnings are available via JDBC  ...

This obviously means that any Tableau (/ any other BI tool) user will not see them. 

> This is meant to convey that the construct has no semantic effect, so users shouldn’t rely on it. 

The construct has an effect (although this isn't warranted by the standard). 
"ORDER BY in subquery has no effect" message can be taken as meaning "if I remove this ORDER BY, my query results should not change", but this is not true (for example `SELECT .. FROM (... ORDER BY) LIMIT 1`). When you're already ignoring ORDER BY (feature is on), this is the right wording. When you're not ignoring ORDER BY (feature is off), something like "subquery uses ORDER BY without OFFSET or FETCH/LIMIT" would be more appropriate.

Reply:
Ok, how about "ORDER BY in subquery may have no effect"?

Reply:
> This obviously means that any Tableau (/ any other BI tool) user will not see them.

Sure, but that doesn't mean we shouldn't add a warning. Otherwise, it's a chicken and egg problem: if we don't have any useful warnings, no one has the incentive to expose them. If no one exposes them, we're not willing to add them.

Reply:
I agree. Let's have a warning. We should still think about increasing warnings visibility. For instance, we could write them to server.log too.

> Ok, how about "ORDER BY in subquery may have no effect"?

I see you don't want to have two different messages depending on toggle's state. This one is better. OK.

Reply:
I think it’s overkill, unnecessarily complicated the code (slightly) and scatters the knowledge about whether the flag is set. I like that now the only place that knows about the toggle is wherever the query plan is being assembled.

Reply:
I think it’s overkill, unnecessarily complicated the code (slightly) and scatters the knowledge about whether the flag is set. I like that now the only place that knows about the toggle is wherever the query plan is being assembled.